### PR TITLE
Towards proper usage of TO_REAL

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards making the rewriter preserve types, which is work towards eliminating subtyping.

This makes it so that the arithmetic rewriter does not simply drop `TO_REAL` from all terms, as this may change the type of the Node. Instead, `TO_REAL` is pulled upwards, and can be removed beneath arithmetic relations. TO_REAL is not eliminated if it occurs as a child of an operator not belonging to arithmetic.  For example if `x,y : Int`:

`(= (+ (to_real x) y) 0.0)` ---> `(= (to_real (+ x y)) 0.0)` ---> `(= (+ x y) 0.0)`  ; note this is the same rewritten form as before
`(P (+ (to_real x) y))` ---> `(P (to_real (+ x y)))`.

The one exception is that `to_real` applied to constants is simply dropped (for now), for example:

`(to_real 5)` ---> `5`

where above, a Real term is rewritten to an Int term.  (Fixing this will require further PRs that make integer constants of kind CONST_INTEGER and integral CONST_RATIONAL have Real type, thus we can have the above rewrite return `5` that is constant and has type Real).

Several quantifiers utilities are patched to preserve their behavior wrt handling TO_REAL.

Finally, a few fixes for subtypes are made to the regressions that we were not catching as errors before, and adds one regression.

The net effect of this PR is that the rewritten form of terms may have `to_real` applications that occur as direct children of symbols from other theories. Special care is required for preregistration, shared terms, and getting model values in the linear solver to strip off `TO_REAL` if necessary before using the (unmodified) interface to the linear solver.

FYI @barrettcw

### CODE REVIEW

File: src/theory/arith/arith_rewriter.cpp [hunk 1]
```diff
@@ -975,7 +1030,7 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t)
             if (rem.isNull())
             {
               return RewriteResponse(REWRITE_DONE,
-                                     nm->mkConstReal(Rational(0)));
+                                     ensureReal(nm->mkConstReal(Rational(0))));
```
Comment (nafur) [inline src/theory/arith/arith_rewriter.cpp:1033]: why is this necessary?

Comment (ajreynol) [reply src/theory/arith/arith_rewriter.cpp:1033]: It's not, this was the byproduct of one the many things I was trying :)  I'll revert.

File: src/theory/arith/arith_rewriter.cpp [hunk 2]
```diff
@@ -530,42 +553,68 @@ RewriteResponse ArithRewriter::rewriteDiv(TNode t, bool pre)
...
+          REWRITE_DONE,
+          ensureReal(rewriter::mkConst(left.getConst<Rational>() / den)));
     }
     if (rewriter::isRAN(left))
     {
-      return RewriteResponse(REWRITE_DONE,
-                             rewriter::mkConst(rewriter::getRAN(left) / den));
+      return RewriteResponse(
+          REWRITE_DONE,
+          ensureReal(rewriter::mkConst(rewriter::getRAN(left) / den)));
```
Comment (nafur) [inline src/theory/arith/arith_rewriter.cpp:579]: Actually, it may be the case that `left == den`, in this case the resulting constant is not necessarily a real...

**Review by nafur — APPROVED** [review]
One comment, not sure if it is important, though.
Otherwise LGTM.

File: src/theory/arith/arith_rewriter.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/arith/arith_rewriter.cpp:579]: The type rule for division says that `(/ t s)` is always a real term, so we actually should always ensure Real here I believe.

Comment (nafur) [reply src/theory/arith/arith_rewriter.cpp:579]: Ah right.
